[RDAPI-23220] Adding use of regular expression section to redaction (#1973)
Axway/axway-open-docs: content/en/docs/apim_administration/apigtw_admin/admin_redactors.md

> * `multi`: Specifies whether the match is multi-line. In multi-line mode, `^` matches the beginning of line, and `$` matches the end of line. Defaults to `false`.

Reviewer: ```suggestion
* `multi`: Specifies whether the match is multi-line. In multi-line mode, `^` and `$` match the beginning and end of any line within the redacted text. Defaults to `false`.
```

Minor change to correct what looks like a mistake. The original doesn't make any sense, does it?

* `utf`: Specifies whether the data being parsed is checked for valid UTF-8 characters. Defaults to `true`. `RawRedactor` cannot be used with non-UTF-8 characters.

### Use of regular expressions

> The way regular expression patterns are written can result in performance drop and excessive use of resources (memory, stack size, CPU). An official PCRE ducomentation page is providing technical details ([PCRE PERFORMANCE](https://www.pcre.org/original/doc/html/pcreperform.html)).

Reviewer 2: ```suggestion
Inefficiently written regular expressions can result in poor performance and excessive use of resources such as memory, stack, and CPU.  The official [PCRE documentation page](https://www.pcre.org/original/doc/html/pcreperform.html) contains suggestions for how to make sure your regular expressions are written correctly.
```

> Some basic guidance rules to write efficient regular expressions could be:

Reviewer 2: ```suggestion
To make your regular expressions more efficient, you should:
```

> * Avoid using excessive recursion ([PCRE RECURSIVE PATTERNS](https://www.pcre.org/original/doc/html/pcrepattern.html#SEC23)) and back references ([PCRE BACK REFERENCES](https://www.pcre.org/original/doc/html/pcrepattern.html#SEC19)).

Reviewer 2: ```suggestion
* Avoid using [excessive recursion](https://www.pcre.org/original/doc/html/pcrepattern.html#SEC23) and [back references](https://www.pcre.org/original/doc/html/pcrepattern.html#SEC19) in your patterns as much as possible.
```

> * Use non-greedy patterns ([PCRE REPETITION](https://www.pcre.org/original/doc/html/pcrepattern.html#SEC17)).

Reviewer 2: ```suggestion
* Use [non-greedy patterns](https://www.pcre.org/original/doc/html/pcrepattern.html#SEC17) when possible.
```

> * Avoid complexity: using different specific patterns for different needs is generally more efficient than using a single huge complex and generic expression.

Reviewer 2: ```suggestion
* Use multiple regex patterns instead of one complex pattern.  Simple, specific patterns are more efficient than combining multiple patterns into a single expression with alternation.  Alternation may consume excessive stack memory.
```

> Use of PCRE recursive pattern is known to be able to cause stack overflows, which leads to process crash. On startup, API Gateway is evaluating the maximum recursion limit to use. That maximum recursion limit is then dynamically lowered whenever stack size is estimated to be too low for the ongoing redaction.

Reviewer 2: ```suggestion
Using recursive patterns may cause stack overflows, which has resulted in crashes. At startup, the API Gateway evaluates the maximum recursion limit to use, which defaults to 100. That maximum recursion limit is dynamically lowered whenever stack size is estimated to be dangerously low in order to prevent crashes.
```

Author: There is no default value, this part is not to be included.

Reviewer 2: I thought the Jira claimed there was a limit of 100?  But if we don't want to discuss that it's fine.
